// @vavite/expose-vite-dev-server@3.0.0/dist/index.js
// src/index.ts
import crypto from "node:crypto";
function vaviteDevServerPlugin() {
  let dev;
  let viteDevServer;
  let globalSymbol;
  function getModuleContents() {
    return `export default ${globalSymbol}`;
  }
  return {
    name: "@vavite/expose-vite-dev-server",
    enforce: "pre",
    buildStart() {
      globalSymbol = "VAVITE_VITE_DEV_SERVER_" + crypto.randomBytes(20).toString("hex");
      global[globalSymbol] = viteDevServer;
    },
    closeBundle() {
      delete global[globalSymbol];
    },
    resolveId(source, _importer, options) {
      if ((source === "@vavite/expose-vite-dev-server/vite-dev-server" || source === "vavite/vite-dev-server" || source === "virtual:vavite-vite-dev-server") && dev && options.ssr) {
        return "virtual:vavite-vite-dev-server";
      }
    },
    load(id, options) {
      if (id === "virtual:vavite-vite-dev-server" && dev && (options == null ? void 0 : options.ssr)) {
        return getModuleContents();
      }
    },
    config(_config, env) {
      dev = env.command === "serve";
      const out = {
        ssr: {
          noExternal: ["vavite"],
          optimizeDeps: {
            exclude: [
              "@vavite/expose-vite-dev-server",
              "virtual:vavite-vite-dev-server"
            ]
          }
        },
        optimizeDeps: {
          exclude: [
            "@vavite/expose-vite-dev-server",
            "virtual:vavite-vite-dev-server"
          ]
        }
      };
      return out;
    },
    configureServer(server) {
      viteDevServer = server;
    }
  };
}
export {
  vaviteDevServerPlugin as default
};
