// @vavite/reloader@3.0.0/dist/index.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  reloader: () => reloader
});
module.exports = __toCommonJS(src_exports);
var import_node_crypto = __toESM(require("crypto"), 1);
function reloader({
  entry = "/server",
  reloadOn = "any-change",
  serveClientAssetsInDev = false
} = {}) {
  let resolvedEntry;
  let entryDeps;
  let globalSymbol;
  function getModuleContents() {
    return `export default ${globalSymbol}`;
  }
  let logger;
  let configEnv;
  let viteServer;
  let resolveListenerPromise;
  const listenerPromise = new Promise(
    (resolve) => resolveListenerPromise = resolve
  );
  async function loadEntry() {
    logger.info("@vavite/reloader: Loading server entry");
    const resolved = await viteServer.pluginContainer.resolveId(
      entry,
      void 0,
      {
        ssr: true
      }
    );
    if (!resolved) {
      logger.error(`@vavite/reloader: Server entry "${entry}" not found`);
      return;
    }
    resolvedEntry = resolved.id;
    await viteServer.ssrLoadModule(resolvedEntry);
    if (reloadOn === "any-change")
      return;
    entryDeps = /* @__PURE__ */ new Set([resolvedEntry]);
    for (const id of entryDeps) {
      const module2 = await viteServer.moduleGraph.getModuleById(id);
      if (!module2) {
        continue;
      }
      if (!module2.ssrTransformResult) {
        continue;
      }
      for (const newDep of module2.ssrTransformResult.deps || []) {
        if (!newDep.startsWith("/")) {
          continue;
        }
        let newId;
        if (newDep.startsWith("/@id/")) {
          newId = newDep.slice(5);
        } else {
          const resolved2 = await viteServer.pluginContainer.resolveId(
            newDep,
            id,
            {
              ssr: true
            }
          );
          if (!resolved2) {
            continue;
          }
          newId = resolved2.id;
        }
        entryDeps.add(newId);
      }
    }
  }
  return {
    name: "@vavite/reloader",
    enforce: "pre",
    buildStart() {
      globalSymbol = "VAVITE_HTTP_DEV_SERVER_" + import_node_crypto.default.randomBytes(20).toString("hex");
    },
    closeBundle() {
      delete global[globalSymbol];
    },
    resolveId(source) {
      if (source === "@vavite/reloader/http-dev-server" || source === "vavite/http-dev-server" || source === "virtual:vavite-http-dev-server") {
        return "virtual:vavite-http-dev-server";
      }
    },
    load(id, options) {
      if (id === "virtual:vavite-http-dev-server") {
        if (!(options == null ? void 0 : options.ssr)) {
          this.error(
            "'vavite/http-dev-server' module is only available in SSR mode"
          );
        }
        return (options == null ? void 0 : options.ssr) && configEnv.command === "serve" ? getModuleContents() : RUNTIME_MODULE_STUB;
      }
    },
    config(config, env) {
      var _a;
      configEnv = env;
      const out = {
        ssr: {
          noExternal: ["vavite"],
          optimizeDeps: {
            exclude: [
              "@vavite/reloader",
              "vavite",
              "virtual:vavite-http-dev-server",
              "vavite/http-dev-server"
            ]
          }
        },
        optimizeDeps: {
          // This silences the "could not auto-determine entry point" warning
          include: [],
          exclude: [
            "@vavite/reloader",
            "vavite",
            "vavite/http-dev-server",
            "virtual:vavite-http-dev-server"
          ]
        }
      };
      if (((_a = config.build) == null ? void 0 : _a.ssr) && env.command === "build") {
        return {
          ...out,
          build: {
            rollupOptions: {
              input: { index: entry }
            }
          }
        };
      }
      return out;
    },
    configResolved(config) {
      logger = config.logger;
    },
    configureServer(server) {
      var _a;
      viteServer = server;
      let listener;
      function addMiddleware() {
        server.middlewares.use(async (req, res) => {
          function renderError(status, message) {
            res.statusCode = status;
            res.end(message);
          }
          await listenerPromise;
          req.url = req.originalUrl;
          try {
            await listener(req, res, () => {
              if (!res.writableEnded)
                renderError(404, "Not found");
            });
          } catch (err) {
            if (err instanceof Error) {
              server.ssrFixStacktrace(err);
              return renderError(500, err.stack || err.message);
            } else {
              return renderError(500, "Unknown error");
            }
          }
        });
      }
      (_a = viteServer.httpServer) == null ? void 0 : _a.on("listening", () => {
        global[globalSymbol] = new Proxy(server.httpServer, {
          get(target, prop) {
            if (prop === "addListener" || prop === "on") {
              return (event, ...rest) => {
                if (event === "request") {
                  listener = rest[0];
                  resolveListenerPromise();
                } else {
                  return target[prop].apply(target, [event, ...rest]);
                }
              };
            } else if (prop === "listen") {
              return (...args) => {
                const listener2 = args.find((arg) => typeof arg === "function");
                if (listener2)
                  Promise.resolve().then(listener2);
                resolveListenerPromise();
              };
            }
            return target[prop];
          }
        });
        loadEntry();
      });
      if (serveClientAssetsInDev) {
        return addMiddleware;
      } else {
        addMiddleware();
      }
    },
    async handleHotUpdate(ctx) {
      if (reloadOn === "any-change") {
        await loadEntry();
        return;
      }
      if (ctx.modules.some((module2) => module2.id && entryDeps.has(module2.id))) {
        await loadEntry();
      }
    }
  };
}
var RUNTIME_MODULE_STUB = `export default undefined`;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  reloader
});
